components/proto/src/lib.rs

```diff
@@ -0,0 +1,293 @@
+    T: ProtobufConvert,
+    S: Default + std::hash::BuildHasher,
+{
+    type ProtoStruct = HashMap<K, T::ProtoStruct, S>;
+    fn to_pb(&self) -> Self::ProtoStruct {
+        self.iter().map(|(k, v)| (k.clone(), v.to_pb())).collect()
+    }
+    fn from_pb(mut pb: Self::ProtoStruct) -> Result<Self, failure::Error> {
+        pb.drain()
+            .map(|(k, v)| ProtobufConvert::from_pb(v).map(|v| (k, v)))
+            .collect::<Result<HashMap<_, _, _>, _>>()
+    }
+}
+
+impl ProtobufConvert for bit_vec::BitVec {
+    type ProtoStruct = BitVec;
+
+    fn to_pb(&self) -> BitVec {
+        let mut bit_vec = BitVec::new();
+        bit_vec.set_data(self.to_bytes());
+        bit_vec.set_len(self.len() as u64);
+        bit_vec
+    }
+
+    fn from_pb(pb: BitVec) -> Result<Self, Error> {
+        let data = pb.get_data();
+        let mut bit_vec = bit_vec::BitVec::from_bytes(data);
+        bit_vec.truncate(pb.get_len() as usize);
+        Ok(bit_vec)
+    }
+}
+
+macro_rules! impl_protobuf_convert_scalar {
+    ($name:tt) => {
+        impl ProtobufConvert for $name {
+            type ProtoStruct = $name;
+            fn to_pb(&self) -> Self::ProtoStruct {
+                *self
+            }
+            fn from_pb(pb: Self::ProtoStruct) -> Result<Self, Error> {
+                Ok(pb)
+            }
+        }
+    };
+}
+
+impl_protobuf_convert_scalar!(bool);
+impl_protobuf_convert_scalar!(u32);
+impl_protobuf_convert_scalar!(u64);
+impl_protobuf_convert_scalar!(i32);
+impl_protobuf_convert_scalar!(i64);
+impl_protobuf_convert_scalar!(f32);
+impl_protobuf_convert_scalar!(f64);
+
+macro_rules! impl_protobuf_convert_fixed_byte_array {
+    ( $( $arr_len:expr ),* ) => {
+        $(
+            /// Special case for fixed sized arrays.
+            impl ProtobufConvert for [u8; $arr_len] {
+                type ProtoStruct = Vec<u8>;
+
+                fn to_pb(&self) -> Self::ProtoStruct {
+                    self.to_vec()
+                }
+
+                fn from_pb(pb: Self::ProtoStruct) -> Result<Self, Error> {
+                    ensure!(
+                        pb.len() == $arr_len,
+                        "wrong array size: actual {}, expected {}",
+                        pb.len(),
+                        $arr_len
+                    );
+
+                    Ok({
+                        let mut array = [0; $arr_len];
+                        array.copy_from_slice(&pb);
+                        array
+                    })
+                }
+            }
+        )*
+    };
+}
+
+// We implement array conversion only for most common array sizes that uses
+// for example in cryptography.
+impl_protobuf_convert_fixed_byte_array! {
+    8, 16, 24, 32, 40, 48, 56, 64,
+    72, 80, 88, 96, 104, 112, 120, 128,
+    160, 256, 512, 1024, 2048
+}
+
+// Think about bincode instead of protobuf. [ECR-3222]
+#[macro_export]
+macro_rules! impl_binary_key_for_binary_value {
+    ($type:ident) => {
```

Comment:
It seems that here is more useful use the `$ty` maybe, so can you check macros in expressions like bellow?

```rust
impl_binary_key_for_binary_value! { foo::Bar }
```